Update PSE 8 with new question flow (#64)
Ada-Developers-Academy/core-pses: 08/kth-missing-positive-number.md

<br>

In the box below, please submit:
1. A shareable link to your conversation in ChatGPT
    - [Documentation for creating a shareable link in ChatGPT](https://help.openai.com/en/articles/7925741-chatgpt-shared-links-faq)
2. Your reflections and summary of the discussion with ChatGPT

##### !end-question
##### !hint

**Troubleshooting**
- If you are having issues with the tool understanding the prompt, try formatting the problem statement or examples differently.
- If you’ve reformatted the information and are still not getting useful results, reach out in #study-hall and share what you are experiencing and the link to your chat so folks can take a look and help you troubleshoot!

<br>

**Summarizing the Review**
- Did the AI tool uncover anything about the observations you made that you hadn’t considered?
- Did the AI tool uncover anything about the questions you asked that you hadn’t considered?
- Did the AI tool suggest updates to the observations you made or questions you asked? 
    - If so, what updates and why?
- Did the AI tool suggest any new observations or questions?
    - If so, what? Why would they be useful?

##### !end-hint
##### !explanation 

> For an example of what a review response might look like, let’s say that we used the example response from the "Explanation" section of the previous question to complete the review prompt. 

Reviewer: Similar update from PSE 7 to give us a little more wiggle room.
```suggestion
For an example of what a review response might look like, let’s say that we provided observations similar to the example response from the "Explanation" section of the previous question to complete the review prompt. 
```

    * Consider at least one nominal and one edge case.
    * What is the expected output for the given input?
    * You can use the examples provided in the prompt, or other examples.

> 2. Write unit tests for `find_missing_positive_number` for the nominal and edge cases you identified in the first step.

Reviewer: Regex replacement of "find_missing_positive_number" to "find_kth_missing_positive_number" worked fine in this file.

- The objective is to create a roadmap that we can use to keep ourselves oriented towards our goal
- It is okay to leave some of the finer details to be worked out in the implementation itself!

As you write your steps, keep the following guidelines in mind:
* We want to think about a general approach rather than what the code would look like line-by-line. 
* It may be helpful to break up the problem/algorithm into smaller subproblems/algorithms. 
    * For example: 1. Handle edge cases, 2. Perform the computation/solve the problem/etc.
* The steps should be a description as if you were talking out the problem with another person and should be agnostic of any particular language. 
    * As such, they should not include code syntax in the description.

What's important at this stage is to think through and outline the implementation before writing code.

##### !end-question
##### !placeholder

Write the logical steps here.

##### !end-placeholder
### !explanation

Example Steps for an O(n) solution:

1. Handle edge case:

>     * If `kth_missing` is smaller than the first value in `numbers`, then `kth_missing` is itself the missing number. Return `kth_missing`.

Reviewer: Edge case for empty list, which out current linear time implementation would fail for.
```suggestion
    * If `numbers` is empty, then `kth_missing` is itself the missing number. Return `kth_missing`.
        * Since every number is missing, and the numbers and `kth_missing` both start from `1`, `kth_missing` is the missing number.
    * If `kth_missing` is smaller than the first value in `numbers`, then `kth_missing` is itself the missing number. Return `kth_missing`.
```

>         * This is becasue it would take more numbers than `kth_missing` to even make us look at the array values. 

Reviewer: ```suggestion
        * This is because it would take more numbers than `kth_missing` to even make us look at the array values. 
```

- The objective is to create a roadmap that we can use to keep ourselves oriented towards our goal
- It is okay to leave some of the finer details to be worked out in the implementation itself!

As you write your steps, keep the following guidelines in mind:
* We want to think about a general approach rather than what the code would look like line-by-line. 
* It may be helpful to break up the problem/algorithm into smaller subproblems/algorithms. 
    * For example: 1. Handle edge cases, 2. Perform the computation/solve the problem/etc.
* The steps should be a description as if you were talking out the problem with another person and should be agnostic of any particular language. 
    * As such, they should not include code syntax in the description.

What's important at this stage is to think through and outline the implementation before writing code.

##### !end-question
##### !placeholder

Write the logical steps here.

##### !end-placeholder
### !explanation

Example Steps for an O(n) solution:

> 1. Handle edge case:

Reviewer: The extra content I added below is probably a bit wordy and could still be reworked, but what I'd like to see is more of a description of why all of this works. The meaning of the adjustment to `missing_target` feels non-obvious to me, so most of what I added is around making sense of that. But overall, for most of the places where we have a step like "if this, do that", I'd like it to be more of the form "if this, that means the following, therefore we do that."

   * If `missing_target` is smaller than the first value in `numbers`, return `missing_target`
2. Create a loop over the input `numbers` that: 
   * starts at the first element of `numbers` 

>    * ends one element before the last element of `numbers`

Reviewer: ```suggestion
2. Loop over the input `numbers`: 
   * start at the first element of `numbers` 
   * end one element before the last element of `numbers`
```

3. Inside the loop:
    1. Check if there are numbers missing between the current element and the next element in `numbers`. If there are missing numbers, store the count in a variable.

>     2. If `missing_target` is less than or equal to the count variable just created, return the value of `numbers` at the current index plus `missing_target` 

Reviewer: ```suggestion
    2. If `missing_target` is less than or equal to the gap size variable just created, the missing number is found within this gap. `missing_target` has been reduced with each gap we've found, so what's left is the number of values beyond the value at the current index. Return the value of `numbers` at the current index plus `missing_target` (the remaining count of missing values).
```

    3. If we did not return, update the variable `missing_target` by subtracting the new missing count variable

> 4. If the loop ends, return the last element of the list plus the remaining value of `missing_target`.

Reviewer: ```suggestion
4. If the loop ends, this means that even accounting for all of the gaps in the array, there are still numbers left to find. `missing_target` has been reduced by the total size of the gaps found in the array, so the missing value we're looking for is the final value in the array plus the remaining count of missing values `missing_target`. Return the last element of the list plus the remaining value of `missing_target`.
```
